### PR TITLE
add asset_key property to context

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Adds an `asset_key` property to `OpExectutionContext`/`AssetExecutionContext` 

## How I Tested These Changes
updated existing usages of `context.asset_key_for_output()` to `context.asset_key`

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16268** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16268" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16268?utm_source=stack-comment).

**Comment by schrockn** [comment]
What is the error message if it is a `multi_asset`?

**Comment by jamiedemaria** [comment]
it'll fall back on `check.failed(f"Output '{output_name}' has no asset")` in `asset_key_for_output`. so it'll say `Output 'result' has no asset` which is not great. I could add a check before calling `asset_key_for_output` and raise a specific error if it's a multi asset

**Comment by schrockn** [comment]
Yeah let's do a specific error message

File: python_modules/dagster/dagster/_core/execution/context/compute.py [hunk 1]
```diff
@@ -446,6 +446,16 @@ def get_mapping_key(self) -> Optional[str]:
     # asset related methods
     #############################################################################################
 
+    @public
+    @property
+    def asset_key(self) -> AssetKey:
+        if self.has_assets_def and len(self.assets_def.keys_by_output_name.keys()) > 1:
+            check.failed(
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/execution/context/compute.py:453]: nit: prefer `raise DagsterInvariantViolation` over check error

**Review by schrockn — CHANGES_REQUESTED** [review]
Yeah what Alex said. check generally should be thought as the moral equivalent of a assert or type failure, rather than this which is more about business logic.

Please add test to verify error behavior then this looks gtg.

**Comment by jamiedemaria** [comment]
where do context tests live?

**Comment by schrockn** [comment]
> where do context tests live?

Don't know off the top of my head! I think I would just grep around the codebase and find the most reasonable spot I could.

**Comment by alangenfeld** [comment]
>  where do context tests live?

personally dont think its terribly important where you put the test, so what ever you prefer. I've recently used `python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py` 

I think its worth testing both `execute_in_process` / `materialize` and direct invocation since they use different context impls and its worth making sure it works in both

**Comment by jamiedemaria** [comment]
yeah i wasn't finding a place where context methods were getting tested, hence the question. I'll just throw it in `test_assets` then

File: python_modules/dagster/dagster/_core/execution/context/compute.py [hunk 2]
```diff
@@ -446,6 +446,17 @@ def get_mapping_key(self) -> Optional[str]:
     # asset related methods
     #############################################################################################
 
+    @public
+    @property
+    def asset_key(self) -> AssetKey:
+        """The AssetKey for the current asset. In a multi_asset, use asset_key_for_output instead."""
+        if self.has_assets_def and len(self.assets_def.keys_by_output_name.keys()) > 1:
+            raise DagsterInvariantViolationError(
+                "Cannot call `context.asset_key` in a multi_asset. Use"
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/execution/context/compute.py:455]: "Cannot call `context.asset_key` in a multi_asset that materializes more than one asset"

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py [hunk 3]
```diff
@@ -1995,3 +1995,29 @@ def use_deps():
...
+        outs={
+            "my_string_asset": AssetOut(),
+            "my_int_asset": AssetOut(),
+        }
+    )
+    def asset_key_context(context: AssetExecutionContext):
+        return context.asset_key, 12
+
+    with pytest.raises(DagsterInvariantViolationError):
+        materialize([asset_key_context])
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_assets.py:2023]: Can you do a string match for "Cannot call `context.asset_key` in a multi_asset" in the test case

**Review by schrockn — APPROVED** [review]
Please heed final comments. Thanks!

**Pull request merged**